File: minecraft-web-manager/old-nodejs-system/backend/src/services/server.service.js
import fs from 'fs/promises';
import path from 'path';
import BashService from './bash.service.js';

class ServerService {
  constructor() {
    this.serverPath = process.env.SERVER_PATH;
    this.bashService = new BashService(this.serverPath);
  }

  /**
   * Obtener estado del servidor
   */
  async getStatus() {
    try {
      const pidPath = path.join(this.serverPath, 'server.pid');

      // Verificar si existe el archivo PID
      try {
        const pidContent = await fs.readFile(pidPath, 'utf-8');
        const pid = parseInt(pidContent.trim());

        // Verificar si el proceso está corriendo
        try {
          process.kill(pid, 0); // Signal 0 solo verifica si el proceso existe

          // Obtener información del proceso
          const { stdout: memInfo } = await this.bashService.execCommand(
            `ps -p ${pid} -o rss,vsz,pcpu --no-headers`
          );
          const [rss, vsz, cpu] = memInfo.trim().split(/\s+/);

          // Obtener uptime
          const { stdout: startTime } = await this.bashService.execCommand(
            `ps -p ${pid} -o lstart --no-headers`
          );
          const startDate = new Date(startTime.trim());
          const uptimeMs = Date.now() - startDate.getTime();
          const uptimeSeconds = Math.floor(uptimeMs / 1000); // Convertir a segundos

          // Intentar obtener jugadores via RCON
          let players = { online: 0, max: 20 };
          try {
            const rconResult = await this.bashService.rconCommand('list');
            const match = rconResult.stdout.match(/There are (\d+) of a max of (\d+) players online/);
            if (match) {
              players = {
                online: parseInt(match[1]),
                max: parseInt(match[2])
              };
            }
          } catch (error) {
            // RCON no disponible o servidor no responde
          }

          // Memoria: convertir de KB a bytes para el frontend
          const rssKB = parseInt(rss);
          const vszKB = parseInt(vsz);

          return {
            running: true,
            pid,
            memory: {
              used: rssKB * 1024, // Convertir KB a bytes
              max: vszKB * 1024    // Convertir KB a bytes
            },
            cpu: parseFloat(cpu),
            uptime: uptimeSeconds, // En segundos
            players
          };
        } catch (error) {
          // Proceso no existe, archivo PID obsoleto
          await fs.unlink(pidPath).catch(() => {});
          return { running: false };
        }
      } catch (error) {
        // Archivo PID no existe
        return { running: false };
      }
    } catch (error) {
      console.error('Error al obtener estado:', error);
      throw error;
    }
  }

  /**
   * Iniciar servidor
   */
  async start() {
    try {
      const status = await this.getStatus();
      if (status.running) {
        throw new Error('El servidor ya está corriendo');
      }

      const result = await this.bashService.serverCommand('start');

      // Esperar un momento para que el servidor inicie
      await new Promise(resolve => setTimeout(resolve, 2000));

      return {
        success: true,
        message: 'Servidor iniciado correctamente',
        output: result.stdout
      };
    } catch (error) {
      throw new Error(`Error al iniciar servidor: ${error.message}`);
    }
  }

  /**
   * Detener servidor
   */
  async stop() {
    try {
      const status = await this.getStatus();
      if (!status.running) {
        throw new Error('El servidor no está corriendo');
      }

      const result = await this.bashService.serverCommand('stop');

      return {
        success: true,
        message: 'Servidor detenido correctamente',
        output: result.stdout
      };
    } catch (error) {
      throw new Error(`Error al detener servidor: ${error.message}`);
    }
  }

  /**
   * Reiniciar servidor
   */
  async restart() {
    try {
      const status = await this.getStatus();
      if (!status.running) {
        // Si no está corriendo, solo iniciarlo
        return await this.start();
      }

      const result = await this.bashService.serverCommand('restart');

      // Esperar un momento para que el servidor reinicie
      await new Promise(resolve => setTimeout(resolve, 3000));

      return {
        success: true,
        message: 'Servidor reiniciado correctamente',
        output: result.stdout
      };
    } catch (error) {
      throw new Error(`Error al reiniciar servidor: ${error.message}`);
    }
  }

  /**
   * Obtener logs del servidor
   */
  async getLogs(lines = 100) {
    try {
      const logsPath = path.join(this.serverPath, 'logs/latest.log');
      const { stdout } = await this.bashService.execCommand(
        `tail -n ${lines} "${logsPath}"`
      );
      return stdout;
    } catch (error) {
      throw new Error(`Error al obtener logs: ${error.message}`);
    }
  }

  /**
   * Obtener información del servidor
   */
  async getInfo() {
    try {
      // Leer versión de Paper
      const versionHistoryPath = path.join(this.serverPath, 'version_history.json');
      let paperVersion = 'Desconocida';
      try {
        const versionData = await fs.readFile(versionHistoryPath, 'utf-8');
        const versionHistory = JSON.parse(versionData);
        if (versionHistory.currentVersion) {
          paperVersion = versionHistory.currentVersion;
        }
      } catch (error) {
        // Archivo no existe o es inválido
      }

      // Obtener versión de Java
      const { stdout: javaVersion } = await this.bashService.execCommand('java -version 2>&1 | head -n 1');

      // Leer server.properties
      const propsPath = path.join(this.serverPath, 'server.properties');
      let serverPort = 25565;
      let maxPlayers = 20;
      try {
        const propsContent = await fs.readFile(propsPath, 'utf-8');
        const portMatch = propsContent.match(/server-port=(\d+)/);
        const maxPlayersMatch = propsContent.match(/max-players=(\d+)/);
        if (portMatch) serverPort = parseInt(portMatch[1]);
        if (maxPlayersMatch) maxPlayers = parseInt(maxPlayersMatch[1]);
      } catch (error) {
        // Usar valores por defecto
      }

      return {
        paperVersion,
        javaVersion: javaVersion.trim(),
        serverPort,
        maxPlayers,
        serverPath: this.serverPath
      };
    } catch (error) {
      throw new Error(`Error al obtener información: ${error.message}`);
    }
  }

  /**
   * Enviar comando via RCON
   */
  async sendCommand(command) {
    try {
      const status = await this.getStatus();
      if (!status.running) {
        throw new Error('El servidor no está corriendo');
      }

      const result = await this.bashService.rconCommand(command);
      return {
        success: true,
        output: result.stdout.trim()
      };
    } catch (error) {
      throw new Error(`Error al enviar comando: ${error.message}`);
    }
  }
}

export default ServerService;

File: minecraft-web-manager/backend-python/app/services/server_service.py
"""Servicio para control del servidor Minecraft"""
import json
import psutil
from pathlib import Path
from typing import Dict, Any, Optional
from app.core.config import settings
from app.services.bash_service import bash_service


class ServerService:
    """Servicio para controlar el servidor Minecraft"""
    
    def __init__(self):
        self.server_path = Path(settings.SERVER_PATH)
        self.pid_file = self.server_path / "server.pid"
    
    def get_pid(self) -> Optional[int]:
        """Obtener PID del servidor desde archivo"""
        try:
            if self.pid_file.exists():
                pid = int(self.pid_file.read_text().strip())
                # Verificar que el proceso existe
                if psutil.pid_exists(pid):
                    return pid
        except Exception:
            pass
        return None
    
    def is_running(self) -> bool:
        """Verificar si el servidor está corriendo"""
        pid = self.get_pid()
        if pid:
            try:
                process = psutil.Process(pid)
                return process.is_running() and process.name() == "java"
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                pass
        return False
    
    def get_version(self) -> Dict[str, str]:
        """
        Obtener versión de Paper del servidor
        
        Returns:
            Dict con paper, minecraft, full
        """
        version_file = self.server_path / "version_history.json"
        
        try:
            if version_file.exists():
                with open(version_file, 'r') as f:
                    data = json.load(f)
                    current_version = data.get('currentVersion', '')
                    
                    # Parsear "1.21.8-60-29c8822 (MC: 1.21.8)"
                    if current_version:
                        # Extraer versión Paper
                        paper_version = current_version.split('-')[0] if '-' in current_version else current_version.split()[0]
                        
                        # Extraer versión Minecraft
                        mc_version = ''
                        if 'MC:' in current_version:
                            mc_version = current_version.split('MC:')[1].strip().rstrip(')')
                        
                        return {
                            "paper": paper_version,
                            "minecraft": mc_version,
                            "full": current_version
                        }
        except Exception as e:
            print(f"Error leyendo versión: {e}")
        
        return {"paper": "Desconocida", "minecraft": "Desconocida", "full": "Desconocida"}
    
    async def get_status(self) -> Dict[str, Any]:
        """
        Obtener estado completo del servidor
        
        Returns:
            Dict con running, pid, memory, cpu, uptime, players
        """
        pid = self.get_pid()
        running = self.is_running()
        
        status = {
            "running": running,
            "pid": pid,
            "memory": {"used": 0, "max": 0},
            "cpu": 0.0,
            "uptime": 0,
            "players": {"online": 0, "max": 20},
            "version": self.get_version()
        }
        
        if running and pid:
            try:
                process = psutil.Process(pid)
                
                # Memoria
                mem_info = process.memory_info()
                status["memory"]["used"] = mem_info.rss
                
                # CPU
                status["cpu"] = process.cpu_percent(interval=0.1)
                
                # Uptime
                status["uptime"] = int(psutil.boot_time() - process.create_time())
                
                # Jugadores (via RCON - solo si mcrcon disponible)
                try:
                    # Verificar si mcrcon está instalado antes de intentar usarlo
                    import shutil
                    if shutil.which("mcrcon"):
                        result = await bash_service.rcon_command("list")
                        if result["success"]:
                            # Parsear "There are X of Y players online"
                            output = result["stdout"]
                            if "of" in output:
                                parts = output.split()
                                status["players"]["online"] = int(parts[2])
                                status["players"]["max"] = int(parts[4])
                except Exception:
                    pass
                
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                status["running"] = False
        
        return status
    
    async def start_server(self) -> Dict[str, Any]:
        """
        Iniciar servidor Minecraft
        
        Returns:
            Dict con success y mensaje
        """
        if self.is_running():
            return {"success": False, "message": "El servidor ya está corriendo"}
        
        result = await bash_service.server_command("start")
        
        return {
            "success": result["success"],
            "message": result["stdout"] if result["success"] else result["stderr"]
        }
    
    async def stop_server(self) -> Dict[str, Any]:
        """
        Detener servidor Minecraft
        
        Returns:
            Dict con success y mensaje
        """
        if not self.is_running():
            return {"success": False, "message": "El servidor no está corriendo"}
        
        result = await bash_service.server_command("stop")
        
        return {
            "success": result["success"],
            "message": result["stdout"] if result["success"] else result["stderr"]
        }
    
    async def restart_server(self) -> Dict[str, Any]:
        """
        Reiniciar servidor Minecraft
        
        Returns:
            Dict con success y mensaje
        """
        result = await bash_service.server_command("restart")
        
        return {
            "success": result["success"],
            "message": result["stdout"] if result["success"] else result["stderr"]
        }
    
    async def get_logs(self, lines: int = 100) -> Dict[str, Any]:
        """
        Obtener últimas líneas de logs
        
        Args:
            lines: Número de líneas
        
        Returns:
            Dict con success y logs
        """
        log_file = self.server_path / "logs" / "latest.log"
        
        try:
            if log_file.exists():
                result = await bash_service.exec_command(
                    f"tail -n {lines} {log_file}",
                    cwd=self.server_path
                )
                return {
                    "success": True,
                    "logs": result["stdout"]
                }
            else:
                return {"success": False, "logs": ""}
        except Exception as e:
            return {"success": False, "logs": "", "error": str(e)}
    
    async def send_command(self, command: str) -> Dict[str, Any]:
        """
        Enviar comando RCON al servidor
        
        Args:
            command: Comando a enviar
        
        Returns:
            Dict con success y respuesta
        """
        if not self.is_running():
            return {"success": False, "message": "El servidor no está corriendo"}
        
        result = await bash_service.rcon_command(command)
        
        return {
            "success": result["success"],
            "output": result["stdout"],
            "error": result["stderr"] if not result["success"] else None
        }


# Instancia global
server_service = ServerService()
